collector/spaces.go

```diff
@@ -0,0 +1,131 @@
+	logger          log.Logger
+	errors          *prometheus.CounterVec
+	client          *godo.Client
+	timeout         time.Duration
+	accessKeyID     string
+	accessKeySecret string
+	Active          *prometheus.Desc
+}
+
+// Templated since each region has a different endpoint
+const spacesDomain = "%s.digitaloceanspaces.com"
+
+// SpacesCollector returns a new SpacesCollector.
+func NewSpacesCollector(logger log.Logger, errors *prometheus.CounterVec, client *godo.Client, accessKeyID string, accessKeySecret string, timeout time.Duration) *SpacesCollector {
+	errors.WithLabelValues("spaces_bucket").Add(0)
+
+	labels := []string{"creation_date", "region", "name"}
+	return &SpacesCollector{
+		logger:          logger,
+		errors:          errors,
+		client:          client,
+		timeout:         timeout,
+		accessKeyID:     accessKeyID,
+		accessKeySecret: accessKeySecret,
+
+		Active: prometheus.NewDesc(
+			"digitalocean_spaces_bucket",
+			"Spaces bucket and its details. Will always be 1 if exists",
+			labels, nil,
+		),
+	}
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector.
+func (c *SpacesCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.Active
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+func (c *SpacesCollector) Collect(ch chan<- prometheus.Metric) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+
+	regions, _, err := c.client.Regions.List(ctx, nil)
+	if err != nil {
+		c.errors.WithLabelValues("spaces_bucket").Add(1)
+		level.Warn(c.logger).Log(
+			"msg", "can't list regions",
+			"err", err,
+		)
+		return
+	}
+
+	// The spaces API can be slow when checking each region 1 by 1, speed up by running them concurrently
+	wg := sync.WaitGroup{}
+	defer wg.Wait()
+
+	for _, region := range regions {
+		wg.Add(1)
+		go func(region godo.Region) {
+			defer wg.Done()
+			spacesEndpoint := fmt.Sprintf(spacesDomain, region.Slug)
+
+			spacesClient, err := minio.New(spacesEndpoint, &minio.Options{
+				Creds:  credentials.NewStaticV4(c.accessKeyID, c.accessKeySecret, ""),
+				Secure: true,
+			})
+			if err != nil {
+				c.errors.WithLabelValues("spaces_bucket").Add(1)
+				level.Warn(c.logger).Log(
+					"msg", "can't create minio client",
+					"err", err,
+				)
+				return
+			}
+
+			// Use a separate context than the godo client. The spaces API can be a bit slow
+			buckets, err := spacesClient.ListBuckets(context.Background())
+			if err != nil {
+				// Not all regions may support spaces
+				// Let's not log all of the known failures
+				var dnsError *net.DNSError
+				if errors.As(err, &dnsError) {
+					return
+				}
+				c.errors.WithLabelValues("spaces_bucket").Add(1)
+				level.Warn(c.logger).Log(
+					"msg", "can't list spaces buckets",
+					"err", err,
+				)
+				return
+			}
+
+			for _, bucket := range buckets {
+				labels := []string{
+					bucket.CreationDate.Format(time.RFC3339),
```

Comment:
This is probably not a good idea. Instead, we could have a separate metric where the creation timestamp is the value?

Reply:
That's fair. What do you think about `float64(bucket.CreationDate.Unix())` ? @metalmatze